nengo_viz/components/netgraph.py

```diff
@@ -194,12 +195,44 @@ def message(self, msg):
             print('invalid message', repr(msg))
             return
         action = info.get('act', None)
+        undo = info.get('undo', None)
         if action is not None:
             del info['act']
-            getattr(self, 'act_' + action)(**info)
+            if action in ('auto_expand', 'auto_collapse'):
+                getattr(self, 'act_' + action[5:])(**info)
+            elif action in ('pan', 'zoom', 'create_modal'):
+                # These should not use the undo stack
+                getattr(self, 'act_' + action)(**info)
+            else:
+                act = create_action(action, self, **info)
+                self.viz.undo_stack.append([act])
+                del self.viz.redo_stack[:]
+        elif undo is not None:
+            if undo == '1':
+                self.undo()
+            else:
+                self.redo()
         else:
             print('received message', msg)
 
+    def undo(self):
+        if self.viz.undo_stack:
+            action = self.viz.undo_stack.pop()
+            re = []
+            for act in action:
```

Comment:
Can there ever be more than one action in `action`? It seems to me it's always a list of one. If so, I think we should get rid of the ability to group actions together, since it just makes the code more complex. We can always add it back later if we're going to use it.